// @chainlink/cre-sdk/dist/generated-sdk/capabilities/blockchain/solana/v1alpha/client_sdk_gen.js
import { create, fromJson } from '@bufbuild/protobuf';
import { AccountMetaSchema, ComputeConfigSchema, WriteReportReplySchema, WriteReportRequestSchema, } from '../../../../../generated/capabilities/blockchain/solana/v1alpha/client_pb';
import { ReportResponseSchema, } from '../../../../../generated/sdk/v1alpha/sdk_pb';
import { Report } from '../../../../../sdk/report';
import { hexToBytes } from '../../../../../sdk/utils/hex-utils';
export function x_generatedCodeOnly_wrap_WriteCreReportRequest(input) {
    return {
        remainingAccounts: input.remainingAccounts,
        receiver: input.receiver,
        computeConfig: input.computeConfig,
        report: input.report !== undefined ? new Report(input.report) : undefined,
        $report: true,
    };
}
export function createWriteCreReportRequest(input) {
    return {
        remainingAccounts: (input.remainingAccounts ?? []).map((v) => fromJson(AccountMetaSchema, v)),
        receiver: hexToBytes(input.receiver),
        computeConfig: input.computeConfig !== undefined
            ? fromJson(ComputeConfigSchema, input.computeConfig)
            : undefined,
        report: input.report,
        $report: true,
    };
}
export function x_generatedCodeOnly_unwrap_WriteCreReportRequest(input) {
    return create(WriteReportRequestSchema, {
        remainingAccounts: input.remainingAccounts,
        receiver: input.receiver,
        computeConfig: input.computeConfig,
        report: input.report !== undefined ? input.report.x_generatedCodeOnly_unwrap() : undefined,
    });
}
/**
 * Client Capability
 *
 * Capability ID: solana@1.0.0
 * Capability Name: solana
 * Capability Version: 1.0.0
 */
export class ClientCapability {
    ChainSelector;
    /** The capability ID for this service */
    static CAPABILITY_ID = 'solana@1.0.0';
    static CAPABILITY_NAME = 'solana';
    static CAPABILITY_VERSION = '1.0.0';
    /** Available ChainSelector values */
    static SUPPORTED_CHAIN_SELECTORS = {
        'solana-devnet': 16423721717087811551n,
    };
    constructor(ChainSelector) {
        this.ChainSelector = ChainSelector;
    }
    writeReport(runtime, input) {
        // Handle input conversion - unwrap if it's a wrapped type, convert from JSON if needed
        let payload;
        // Check if it's a wrapped type by looking for the $report property
        if (input.$report) {
            // It's a wrapped type, unwrap it
            payload = x_generatedCodeOnly_unwrap_WriteCreReportRequest(input);
        }
        else {
            // It's wrapped JSON, convert using create function
            payload = x_generatedCodeOnly_unwrap_WriteCreReportRequest(createWriteCreReportRequest(input));
        }
        // Include all labels in capability ID for routing when specified
        const capabilityId = `${ClientCapability.CAPABILITY_NAME}:ChainSelector:${this.ChainSelector}@${ClientCapability.CAPABILITY_VERSION}`;
        const capabilityResponse = runtime.callCapability({
            capabilityId,
            method: 'WriteReport',
            payload,
            inputSchema: WriteReportRequestSchema,
            outputSchema: WriteReportReplySchema,
        });
        return {
            result: () => {
                const result = capabilityResponse.result();
                return result;
            },
        };
    }
}
